pkg/auth/oauth/handlers/default_auth_handler.go

```diff
@@ -210,3 +227,33 @@ func mergeHeaders(dest http.Header, toAdd http.Header) {
 		}
 	}
 }
+
+// getAuthMethods is copied from k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go
+// to be able to return information about failed authentication
+func getAuthMethods(req *http.Request) string {
+	authMethods := []string{}
+
+	if _, _, ok := req.BasicAuth(); ok {
+		authMethods = append(authMethods, "basic")
+	}
+
+	auth := strings.TrimSpace(req.Header.Get("Authorization"))
+	parts := strings.Split(auth, " ")
+	if len(parts) > 1 && strings.ToLower(parts[0]) == "bearer" {
+		authMethods = append(authMethods, "bearer")
+	}
+
+	token := strings.TrimSpace(req.URL.Query().Get("access_token"))
```

Comment:
No error and code params? They are also part of the OAuth flow.

Reply:
That was discussed with @liggitt in the original PR (https://github.com/kubernetes/kubernetes/pull/51119). We want to provide with enough information that auth failed and the method we've tried but without going into too much details.